mteb/encoder_interface.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import Any, Protocol, runtime_checkable
+
+import numpy as np
+import torch
+
+
+@runtime_checkable
+class Encoder(Protocol):
+    """The interface for an encoder in MTEB."""
+
+    def encode(
+        self, sentences: list[str], prompt: str, **kwargs: Any
+    ) -> torch.Tensor | np.ndarray:
+        """Encodes the given sentences using the encoder.
+
+        Args:
+            sentences: The sentences to encode.
+            prompt: The prompt to use. Useful for prompt-based models.
+            **kwargs: Additional arguments to pass to the encoder.
+
+        Returns:
+            The encoded sentences.
+        """
+        ...
+
+
+class EncoderWithQueryCorpusEncode(Encoder, Protocol):
+    """The interface for an encoder that supports encoding a queries and a corpus."""
+
+    def encode_queries(
+        self, queries: list[str], prompt: str, **kwargs: Any
+    ) -> torch.Tensor | np.ndarray:
+        """Encodes the given queries using the encoder.
+
+        Args:
+            queries: The queries to encode.
+            prompt: The prompt to use. Useful for prompt-based models.
+            **kwargs: Additional arguments to pass to the encoder.
+
+        Returns:
+            The encoded queries.
+        """
+        ...
+
+    def encode_corpus(
+        self, corpus: list[str], prompt: str, **kwargs: Any
+    ) -> torch.Tensor | np.ndarray:
+        """Encodes the given corpus using the encoder.
+
+        Args:
+            corpus: The corpus to encode.
```

Comment:
This can also be a dict. Is there a good reason to allow this?

Reply:
I think BEIR loads it as a dict -- may not be relevant, but that's why I've seen it that way.

Reply:
I think we should at least pick a format either dict or str and string would be more in line with existing implementations in MTEB, but a dict would be nice to keep it consistent with beir (I think either is reasonable).

Reply:
Doesn't making this a `str` not allow us to extend it to provide any custom separator tokens? I could be missing something.  I think it's fine for the default to be just concatenating them, but it should be extendable to allow others to provide custom sep tokens

Reply:
Do people use custom separator tokens for title and text?